oak/server/rust/oak_runtime/src/node/logger.rs

```diff
@@ -65,6 +68,9 @@ impl super::Node for LogNode {
                     }
                 },
                 Ok(None) => {}
+                Err(OakStatus::ErrTerminated) | Err(OakStatus::ErrChannelClosed) => {
+                    break;
+                }
```

Comment:
I think the logic in this loop is a bit counter intuitive. We have an explicit termination signal from the runtime, but this node chooses to ignore it because it wants to keep going in case any other node still wants to send a message. This seems to defeat the point of the termination signal, if each node can decide it is more important and ignore it. Should we think of a better design at some point? cc @daviddrysdale

Reply:
I agree it is counter-intuitive, but also that it is needed so that log messages aren't dropped based on the order in which these are created. I can tthink of two options:
- Make logging a host function call rather than a separate node.
- Make sure nodes are stopped in a sequence that will ensure no log messages are dropped. This would probably require keeping track of which nodes created which other nodes and traversing the graph in a specific order when terminating nodes.

Reply:
I don't think the logger should choose to ignore the termination message (I guess this happens to work because all the other Nodes close down first and so release their logging channels, thus triggering the `ErrChannelClosed` exit arm).

However, I've found in the past that it's helpful for the logger to drain pending log messages before shutting down (it makes debugging termination issues much easier).

So I might structure this as two loops, something like:
```Rust
loop {  // Wait for activity
  let wait_result = wait_on_channels();
  loop {  // Drain all currently available log messages
    match channel_read() {
      // ...
      Ok(None) => break,
    }
  }
  if wait_result == ChannelClosed || wait_result == Terminated {
    break;
  }
}
```

(I don't think we want to put anything as a host function that can be implemented as a Node, to reduce API surface and to allow IFC rules to apply straightforwardly.)

Reply:
@conradgrobler could you capture this in an issue, and perhaps look into it next?

I think shutting down nodes in reverse order of creation (or some other well defined order) may be an interesting one to consider.

The nested loop seems like it would help in some cases, but it's not guaranteed, since the inner loop lacks any synchronization (i.e. perhaps another node sends its last message after a 1s delay because it was busy processing some data).

Reply:
Having thought about it more, I don't think the order of the creation helps.

A node could create a log node and then later create another WASM node and send a copy of the write half of the handle used by the log node to the WASM node. Other nodes might create the WASM node first. In both cases we want the log node to receive messages until all write halfs have been closed.

I don't think the double loop code is guaranteed to drain all messages in the case where a the write half of the channel was shared by multiple nodes. It can be fixed by not breaking on `ErrTerminated`, but then it is basically back to the current implementation which will busy-wait during the time the runtime is terminating.

Reply:
I have raised #1100, but set the priority to P2 as it is currently working as intended.